### PR TITLE
Add Toolkit.run

### PR DESCRIPTION
**Pull request by jclem**
**Why?**

I've been finding myself using this pattern over and over:

```js
main().catch(err => {
  tools.log.fatal(err)
  tools.exit.failure()
})

async function main() {
  // Action code
}
```

In order to reduce boilerplate when using async/await, I'm proposing `tools.run`:

```js
Toolkit.run(async tools => {
  // Action code
}, { event: 'push' })
```

**How?**

This adds a simple static `run` function to the Toolkit class. It accepts an async function as an argument, and optionally some toolkit options. If that function throws an error, `run` will log the error and exit with a failure status.

---

- [x] Tests have been added/updated (if necessary)
- [x] Documentation has been updated (if necessary)

### CODE REVIEW

**Comment by JasonEtco** [comment]
Definitely agree with this 👌 I don't like the pattern of a `main` function at all, but I've also had to do something similar in every Action. I have done something else helpful - by having a separate function that is my "action code" that takes a `Toolkit` argument, I can pass a mocked Toolkit in tests. I need to document this for #50, but it vaguely looks like this:

```js
module.exports = async tools => {
  if (!tools) tools = new Toolkit()
}
```

This lets me write an async function and do this in tests:

```js
it('does something', async () => {
  const tools = new Toolkit()
  // Mock something on the Toolkit instance
  tools.getPackageJSON = jest.fn(() => ({ })
  await myAction(tools)
})
```

So with that in mind, what about an API like this:

```ts
class Toolkit {
  static public async run (fn: (tools: Toolkit) => Promise<void>, options: ToolkitOptions) {
    const tools = new Toolkit(options)
    return fn(tools)
  }
}
```

And used like:

```js
const { Toolkit } = require('actions-toolkit')
Toolkit.run(async tools => {
  return tools.github.repos.delete(tools.context.repo)
}, { event: 'issues' })
```

This seems like an easier way to export a function to test, but still run the action code. What do you think @jclem?

**Comment by jclem** [comment]
I think for the pattern of using a single action for many scenarios/actions that makes sense. Personally, I haven't actually done that yet, but I can see why one would.

I think this proposal is great. I can still use it the way I planned to if the `options: ToolkitOptions` argument is optional.

**Comment by JasonEtco** [comment]
> I can still use it the way I planned to if the options: ToolkitOptions argument is optional.

Oh good point, yeah it should definitely still be optional. Also not sure if `=> Promise<void>` is better than `=> Promise<unknown>` or `any` 🤔

**Comment by jclem** [comment]
> Also not sure if => Promise<void> is better than => Promise<unknown> or any 🤔

I *think* `unknown` is right—`void` would probably work, but it's not technically correct (it's not that the callback *must* return no value, it's that we don't care). `unknown` means "we can't use this without a type assertion", which is safer than `any`, where we can use the value if we wanted to.

I wonder if a generic makes more sense, but that seems overcomplicated, since we don't plan on using the value.

**Comment by JasonEtco** [comment]
> unknown means "we can't use this without a type assertion"

Makes sense 👍 🚀 

> I wonder if a generic makes more sense, but that seems overcomplicated, since we don't plan on using the value.

Yeah agreed

File: src/index.ts [hunk 1]
```diff
@@ -16,6 +16,30 @@ export interface ToolkitOptions {
...
+   * Toolkit.run(async tools => {
+   *   // Action code here.
+   * }, { event: 'push' })
+   * ```
+   */
+  public static async run (func: (tools: Toolkit) => Promise<unknown>, opts?: ToolkitOptions) {
+    const tools = new Toolkit(opts)
+
+    return func(tools).catch(err => {
+      tools.log.fatal(err)
```
Comment (JasonEtco) [inline src/index.ts]: `tools.exit.failure` can log a message as it's argument:

https://github.com/JasonEtco/actions-toolkit/blob/07e0eca27fb246116882844edb94ef7a9391debc/src/exit.ts#L37

But now I realize that its not using the logger 😱 I'll get that fixed to use `tools.log.fatal` inside of `tools.exit.failure()`, but it means that we could do:

```suggestion
      tools.exit.failure(err)
```

File: README.md [hunk 2]
```diff
@@ -87,6 +87,19 @@ const tools = new Toolkit({
...
 ```
 
+### Toolkit.run
+
+Run an asynchronous function that receives an instance of `Toolkit` as its argument. If the function throws an error (or returns a rejected promise), `Toolkit.run` will log the error and exit the action with a failure status code.
+
+The toolkit instance can be configured by passing `Toolkit` options as the second argument to `Toolkit.run`.
+
+```js
+Toolkit.run(async tool => {
```
Comment (JasonEtco) [inline README.md]: Tiny nit, just to more consistent with other docs:

```suggestion
Toolkit.run(async tools => {
```

File: tests/index.test.ts [hunk 3]
```diff
@@ -159,6 +159,37 @@ describe('Toolkit', () => {
     })
   })
 
+  describe('.run', () => {
+    it('runs the async function passed to it', async () => {
```
Comment (JasonEtco) [inline tests/index.test.ts]: Could we use a `jest.fn()` mock function here? Or is there some nuance to testing the async function that I'm not picking up on? Like:

```js
const fn = jest.fn()
await Toolkit.run(fn)
expect(fn).toHaveBeenCalled()
expect(fn.mock.calls[0][0]).toBeInstanceOf(Toolkit)
```

**Comment by jclem** [comment]
@JasonEcto No problem at all, I appreciate the help. Got partway through the suggestions, and some things came up. Thanks for wrapping this up!

**Pull request merged**